NBitcoin/Utils.cs

```diff
@@ -322,6 +322,10 @@ public static int ReadEx(this Stream stream, byte[] buffer, int offset, int coun
 
 				int currentReadCount;
 
+#if !NO_SOCKETASYNC
+				currentReadCount = stream.ReadAsync(buffer, offset + totalReadCount, count - totalReadCount, cancellation).GetAwaiter().GetResult();

```

Comment:
I am a bit out of context just saw this code and have this idea, why not use it? 
https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=net-6.0

Reply:
We basically want to be able to cancel that operation. But the layers above are not `async`, so the idea is to use `.GetAwaiter().GetResult()` and maybe in follow-up PRs we could accually make the upper layers async functions (that would require some planning as it would most certainly lead to BC breaks).

Reply:
Oh, I see what is the idea here.

Reply:
The goal of the PR is really to be "as close to .NET 6 as possible" as that means less maintenance and the work is done by .NET team to ensure it works correctly.

Reply:
What ifyou subscribe to the cancellationtoken and close the socket. The idea is that if you Close the socket all the Read operations immediately fail.

Reply:
It may work, yes. However, the general direction of .NET is to use `async` functions and it works great. Using:

> What if you register to the cancellationtoken and close the socket.

does not give us really the hope that NBitcoin will more closer to how people currently write networking code. So personally, I would be in favor of "make more code async gradually" than looking for solutions that fits to the current code base and hide that some parts are async already: https://github.com/MetacoSA/NBitcoin/blob/cb6d64664b9f38d0442a4a4ad5edaed9e310bf40/NBitcoin/Utils.cs#L379